File: implementations/csharp/client/src/ShuaiTunnel.Client/PeerMesh/PeerAppMessageCodec.cs
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ShuaiTunnel.Client.PeerMesh;

internal static class PeerAppMessageCodec
{
    public const string TypeMessage = "message";
    public const string TypeAck = "ack";

    private static readonly byte[] PrefixV1 = Encoding.ASCII.GetBytes("STMSG1\n");
    private static readonly byte[] PrefixV2 = Encoding.ASCII.GetBytes("STMSG2\n");

    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web)
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingDefault,
    };

    public static bool LooksLike(byte[] payload)
    {
        return payload.Length >= PrefixV1.Length
            && (payload.AsSpan(0, PrefixV1.Length).SequenceEqual(PrefixV1)
                || payload.AsSpan(0, PrefixV2.Length).SequenceEqual(PrefixV2));
    }

    public static bool TryDecode(byte[] payload, out PeerAppMessage message)
    {
        message = default!;
        if (!LooksLike(payload))
        {
            return false;
        }
        try
        {
            var decoded = JsonSerializer.Deserialize<PeerAppMessage>(
                payload.AsSpan(PrefixLength(payload)),
                JsonOptions);
            if (decoded is null || string.IsNullOrWhiteSpace(decoded.Type))
            {
                return false;
            }
            message = decoded;
            return true;
        }
        catch (JsonException)
        {
            return false;
        }
    }

    public static byte[] Encode(PeerAppMessage message)
    {
        // Java and Go only understand STMSG1. Keep ordinary text messages and
        // ACKs on that envelope; STMSG2 is solely the optional attachment
        // extension and must never leak attachment metadata into an ACK.
        var attachmentV2 = string.Equals(message.Type, TypeMessage, StringComparison.OrdinalIgnoreCase)
            && message.Attachment is not null;
        var wireMessage = attachmentV2 ? message : WithoutAttachment(message);
        var prefix = attachmentV2 ? PrefixV2 : PrefixV1;
        var json = JsonSerializer.SerializeToUtf8Bytes(wireMessage, JsonOptions);
        var payload = new byte[prefix.Length + json.Length];
        prefix.CopyTo(payload, 0);
        json.CopyTo(payload.AsSpan(prefix.Length));
        return payload;
    }

    public static string DisplayText(PeerAppMessage message)
    {
        if (message.Attachment is null)
        {
            return message.Message ?? "";
        }
        var size = message.Attachment.SizeBytes > 0 ? FormatBytes(message.Attachment.SizeBytes) : "-";
        var text = string.IsNullOrWhiteSpace(message.Message) ? "" : $"{message.Message} ";
        return $"{text}[附件] {FirstNonEmpty(message.Attachment.FileName, message.Attachment.ObjectId, "attachment")} · {FirstNonEmpty(message.Attachment.MimeType, "application/octet-stream")} · {size}";
    }

    private static int PrefixLength(byte[] payload) =>
        payload.AsSpan(0, PrefixV2.Length).SequenceEqual(PrefixV2) ? PrefixV2.Length : PrefixV1.Length;

    private static PeerAppMessage WithoutAttachment(PeerAppMessage message) => new()
    {
        Type = message.Type,
        Id = message.Id,
        FromClientId = message.FromClientId,
        FromClientName = message.FromClientName,
        ToClientId = message.ToClientId,
        ToClientName = message.ToClientName,
        Message = message.Message,
        CreatedAtMillis = message.CreatedAtMillis,
    };

    private static string FormatBytes(long bytes)
    {
        string[] units = ["B", "KB", "MB", "GB", "TB"];
        var value = (double)Math.Max(0L, bytes);
        var unit = 0;
        while (value >= 1024 && unit < units.Length - 1)
        {
            value /= 1024;
            unit++;
        }
        return unit == 0 ? $"{value:0} {units[unit]}" : $"{value:0.##} {units[unit]}";
    }

    private static string FirstNonEmpty(params string?[] values)
    {
        return values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value))?.Trim() ?? "";
    }
}

internal sealed class PeerAppMessage
{
    [JsonPropertyName("type")]
    public string? Type { get; set; }

    [JsonPropertyName("id")]
    public string? Id { get; set; }

    [JsonPropertyName("fromClientId")]
    public long FromClientId { get; set; }

    [JsonPropertyName("fromClientName")]
    public string? FromClientName { get; set; }

    [JsonPropertyName("toClientId")]
    public long ToClientId { get; set; }

    [JsonPropertyName("toClientName")]
    public string? ToClientName { get; set; }

    [JsonPropertyName("message")]
    public string? Message { get; set; }

    [JsonPropertyName("attachment")]
    public PeerAppAttachment? Attachment { get; set; }

    [JsonPropertyName("createdAtMillis")]
    public long CreatedAtMillis { get; set; }
}

internal sealed class PeerAppAttachment
{
    [JsonPropertyName("objectId")]
    public string? ObjectId { get; set; }

    [JsonPropertyName("attachmentId")]
    public long AttachmentId { get; set; }

    [JsonPropertyName("fileName")]
    public string? FileName { get; set; }

    [JsonPropertyName("mimeType")]
    public string? MimeType { get; set; }

    [JsonPropertyName("sizeBytes")]
    public long SizeBytes { get; set; }

    [JsonPropertyName("sha256")]
    public string? Sha256 { get; set; }
}

File: implementations/android/client/app/src/main/java/com/theshuai/tunnel/android/PeerAppMessageCodec.java
package com.theshuai.tunnel.android;

import org.json.JSONObject;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Locale;

final class PeerAppMessageCodec {
    static final String TYPE_MESSAGE = "message";
    static final String TYPE_ACK = "ack";

    private static final byte[] PREFIX = "STMSG2\n".getBytes(StandardCharsets.US_ASCII);

    private PeerAppMessageCodec() {
    }

    static boolean looksLike(byte[] payload) {
        return payload != null
                && payload.length >= PREFIX.length
                && Arrays.equals(PREFIX, Arrays.copyOfRange(payload, 0, PREFIX.length));
    }

    static PeerAppMessage decode(byte[] payload) {
        if (!looksLike(payload)) {
            return null;
        }
        try {
            String jsonText = new String(payload, PREFIX.length, payload.length - PREFIX.length, StandardCharsets.UTF_8);
            JSONObject json = new JSONObject(jsonText);
            String type = json.optString("type", "");
            if (type.trim().isEmpty()) {
                return null;
            }
            PeerAppMessage message = new PeerAppMessage();
            message.type = type;
            message.id = json.optString("id", "");
            message.fromClientId = json.optLong("fromClientId", 0L);
            message.fromClientName = json.optString("fromClientName", "");
            message.toClientId = json.optLong("toClientId", 0L);
            message.toClientName = json.optString("toClientName", "");
            message.message = json.optString("message", "");
            message.attachment = json.optJSONObject("attachment");
            message.createdAtMillis = json.optLong("createdAtMillis", 0L);
            return message;
        } catch (Exception ignored) {
            return null;
        }
    }

    static byte[] encode(PeerAppMessage message) throws Exception {
        boolean attachmentV2 = message != null
                && TYPE_MESSAGE.equalsIgnoreCase(text(message.type))
                && message.attachment != null;
        StringBuilder json = new StringBuilder(256);
        json.append("{\"type\":").append(JSONObject.quote(text(message == null ? null : message.type)))
                .append(",\"id\":").append(JSONObject.quote(text(message == null ? null : message.id)))
                .append(",\"fromClientId\":").append(message == null ? 0L : message.fromClientId)
                .append(",\"fromClientName\":").append(JSONObject.quote(
                        text(message == null ? null : message.fromClientName)))
                .append(",\"toClientId\":").append(message == null ? 0L : message.toClientId)
                .append(",\"toClientName\":").append(JSONObject.quote(
                        text(message == null ? null : message.toClientName)))
                .append(",\"message\":").append(JSONObject.quote(
                        text(message == null ? null : message.message)));
        if (attachmentV2) {
            json.append(",\"attachment\":").append(message.attachment);
        }
        json.append(",\"createdAtMillis\":")
                .append(message == null ? 0L : message.createdAtMillis)
                .append('}');
        byte[] body = json.toString().getBytes(StandardCharsets.UTF_8);
        byte[] payload = new byte[PREFIX.length + body.length];
        System.arraycopy(PREFIX, 0, payload, 0, PREFIX.length);
        System.arraycopy(body, 0, payload, PREFIX.length, body.length);
        return payload;
    }

    static String displayText(PeerAppMessage message) {
        if (message == null) {
            return "";
        }
        if (message.attachment == null) {
            return text(message.message);
        }
        String fileName = firstNonBlank(
                message.attachment.optString("fileName", ""),
                message.attachment.optString("objectId", ""),
                "attachment");
        String mimeType = firstNonBlank(
                message.attachment.optString("mimeType", ""),
                "application/octet-stream");
        long size = message.attachment.optLong("sizeBytes", 0L);
        String prefix = text(message.message).trim();
        return (prefix.isEmpty() ? "" : prefix + " ")
                + "[附件] " + fileName + " · " + mimeType + " · "
                + (size > 0L ? formatBytes(size) : "-");
    }

    private static String text(String value) {
        return value == null ? "" : value;
    }

    private static String firstNonBlank(String... values) {
        if (values != null) {
            for (String value : values) {
                if (value != null && !value.trim().isEmpty()) {
                    return value.trim();
                }
            }
        }
        return "";
    }

    private static String formatBytes(long bytes) {
        String[] units = {"B", "KB", "MB", "GB"};
        double value = Math.max(0L, bytes);
        int unit = 0;
        while (value >= 1024.0 && unit < units.length - 1) {
            value /= 1024.0;
            unit++;
        }
        return String.format(Locale.ROOT, value >= 10 || unit == 0 ? "%.0f %s" : "%.1f %s", value, units[unit]);
    }

    static final class PeerAppMessage {
        String type;
        String id;
        long fromClientId;
        String fromClientName;
        long toClientId;
        String toClientName;
        String message;
        JSONObject attachment;
        long createdAtMillis;
    }
}
